server/src/main/scala/org/http4s/server/middleware/Caching.scala

```diff
@@ -0,0 +1,111 @@
+package org.http4s.server.middleware
+
+import cats._
+import cats.implicits._
+import cats.effect._
+import cats.data._
+import org.http4s._
+import org.http4s.util.CaseInsensitiveString
+import org.http4s.headers.{Date => HDate, _}
+import scala.concurrent.duration._
+
+object Caching {
+  def `no-store`[G[_]: Monad: Clock, F[_], A](
+      http: Kleisli[G, A, Response[F]]): Kleisli[G, A, Response[F]] =
+    Kleisli { a: A =>
+      for {
+        resp <- http(a)
+        now <- HttpDate.current[G]
+      } yield {
+        val headers = List(
+          `Cache-Control`(
+            NonEmptyList.of[CacheDirective](
+              CacheDirective.`no-store`,
+              CacheDirective.`private`(),
+              CacheDirective.`no-cache`(),
+              CacheDirective.`max-age`(0.seconds)
+            )),
+          Header("Pragma", "no-cache"),
+          HDate(now),
+          Expires(HttpDate.Epoch) // Expire at the epoch for no time confusion
+        )
+        resp.putHeaders(headers: _*)
+      }
+    }
+
+  object Helpers {
+    def defaultStatussToSetOn(s: Status): Boolean =
+      s match {
+        case Status.NotModified => true
+        case otherwise => otherwise.isSuccess
+      }
+
+    def defaultMethodsToSetOn(m: Method): Boolean = methodsToSetOn.contains(m)
+
+    private lazy val methodsToSetOn: Set[Method] = Set(
+      Method.GET,
+      Method.HEAD
+    )
+  }
+
+  def publicCache[G[_]: MonadError[*[_], Throwable]: Clock, F[_]](
+      lifetime: Duration,
```

Comment:
I feel like we more commonly put the `Http` alone in a parameter list, for partial application.  But I got that feeling a long time ago, before we had all the `G` and what not.  I'm not sure whether that's still a useful convention.

Reply:
We could put it as the first argument, but if we put it in the second like we generally do, then G and F cannot infer

Reply:
Drat. I think you're right.  That's sad, because I like to think of these as partially applied up to the `Http`.

Reply:
Alternative is building it as partially applied class. So that we can say `Duration ->  (forall g, f . ((MonadError Throwable g), Clock g) => http g f)`

Reply:
Ooh, that's an interesting idea, if the type inference still works well enough when folding them down into one.